plugin-unity-agent/src/main/kotlin/jetbrains/buildServer/unity/PEProductVersionDetector.kt

```diff
@@ -0,0 +1,76 @@
+package jetbrains.buildServer.unity
+
+import com.intellij.openapi.diagnostic.Logger
+import com.vdurmont.semver4j.Semver
+import jetbrains.buildServer.util.PEReader.PEUtil
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.decodeFromString
+import kotlinx.serialization.json.Json
+import java.io.BufferedReader
+import java.io.File
+import java.nio.file.Paths
+import java.util.concurrent.TimeUnit
+import kotlin.io.path.exists
+
+class PEProductVersionDetector {
+    fun detect(executable: File): Semver? {
+        val version = PEUtil.getProductVersion(executable)
+
+        if (version != null) {
+            return Semver("${version.p1}.${version.p2}.${version.p3}", Semver.SemverType.LOOSE)
+        }
+
+        LOG.debug("unable to detect version via PEUtil, falling back to detection tool")
+
+        val detectorPath = Paths.get("../plugins/teamcity-unity-agent/tools/PeProductVersionDetector.exe")
+            .toAbsolutePath()
+
+        if (!detectorPath.exists()) {
+            LOG.debug("PE product version detection tool missing")
+            return null
+        }
+
+        val detector = ProcessBuilder(detectorPath.toString(), executable.absolutePath).start()
+
+        if (detector.waitFor(3, TimeUnit.SECONDS)) {
+            return if (detector.exitValue() == 0) {
+                val versionJson = detector.inputStream
+                    .bufferedReader(Charsets.UTF_8)
+                    .use(BufferedReader::readText)
+
+                val versionObject = Json.decodeFromString<ProductVersion?>(versionJson)
+
+                if (versionObject?.majorPart != null) {
+                    Semver("${versionObject.majorPart}.${versionObject.minorPart}.${versionObject.buildPart}",
+                        Semver.SemverType.LOOSE)
+                } else {
+                    null
+                }
+            } else {
+                val error = detector.errorStream
+                    .bufferedReader(Charsets.UTF_8)
+                    .use(BufferedReader::readText)
+
+                LOG.debug(error)
+                null
+            }
+        } else {
+            detector.destroyForcibly()
+            LOG.debug("PE product version detection timed out")
+            return null
+        }
+    }
+
+    @Serializable
+    private data class ProductVersion(
+        @SerialName("MajorPart") val majorPart: Int?,
+        @SerialName("MinorPart") val minorPart: Int?,
+        @SerialName("BuildPart") val buildPart: Int?,
+        @SerialName("PrivatePart") val privatePart: Int?
+    )
```

Comment:
Just a question: is Unity version always semver?
If they change versioning system will the detection process be able to detect?

Reply:
It's more about the PE product version, the structure is predefined and the detection process will always work and will return numbers (or 0/null if missing).  Only in case, Unity won't put these numbers into [VS_FIXEDFILEINFO](https://learn.microsoft.com/en-us/windows/win32/api/verrsrc/ns-verrsrc-vs_fixedfileinfo) at all, the process will fail